Support for JetBrains Toolbox and JetBrains Fleet editor

tsvetilian-ty opened the pull request:
<!--
What GitHub Desktop issue does this PR address? (for example, #1234)
-->

Closes #12912 #15772 

## Description
- Add support for [JetBrains Toolbox](https://www.jetbrains.com/toolbox-app/)
  - The implementation detects only the default configuration for the scripts
- Add [JetBrains Fleet](https://www.jetbrains.com/fleet/)

### Screenshots
![fleet](https://user-images.githubusercontent.com/9341546/207426925-7a428fc1-dfe2-4f4a-932b-51b2d7d45b31.gif)

tsvetilian-ty committed:
Add support on Windows

tsvetilian-ty committed:
Add JetBrains Fleet for Windows

tsvetilian-ty committed:
Add Fleet as supported editor on darwin systems

tsvetilian-ty committed:
Update docs

tsvetilian-ty committed:
Update type docs

valentinegb commented on app/src/lib/editors/darwin.ts:
```diff
@@ -144,6 +144,10 @@ const editors: IDarwinExternalEditor[] = [
     bundleIdentifiers: ['com.lite-xl'],
   },
+  {
+    name: 'JetBrains Fleet',
```
Other JetBrains IDEs on this list do not have the JetBrains prefix, neither should Fleet. This is only on Mac though, for some reason the JetBrains prefix has been normalized on Windows, apparently.

tsvetilian-ty replied:
You're absolutely right 🙌

tsvetilian-ty committed:
Change JetBrains Fleet to Fleet

Matching the name of JetBrains Fleet with the other JetBrains products in GitHub Desktop on macOS

Kokomo1982 commented:
> Closes #12912 #15772
> 
> ## Description
> * Add support for [JetBrains Toolbox](https://www.jetbrains.com/toolbox-app/)
>   
>   * The implementation detects only the default configuration for the scripts
> * Add [JetBrains Fleet](https://www.jetbrains.com/fleet/)
> 
> ### Screenshots
> ![fleet](https://user-images.githubusercontent.com/9341546/207426925-7a428fc1-dfe2-4f4a-932b-51b2d7d45b31.gif) [ ](https://user-images.githubusercontent.com/9341546/207426925-7a428fc1-dfe2-4f4a-932b-51b2d7d45b31.gif) ![fleet](https://user-images.githubusercontent.com/9341546/207426925-7a428fc1-dfe2-4f4a-932b-51b2d7d45b31.gif) [ ](https://user-images.githubusercontent.com/9341546/207426925-7a428fc1-dfe2-4f4a-932b-51b2d7d45b31.gif)

``

valentinegb reviewed (commented):
This currently only works on Windows.

tsvetilian-ty commented:
Does Fleet show in the dropdown?

valentinegb commented:
It does, but it doesn't use the CLI, it launches the app without opening the specified directory.

tsvetilian-ty commented:
Could you make a video and share it? Because I'm unable to reproduce the issue 🤔

tsvetilian-ty committed:
Add Fleet to the list with supported editors

valentinegb commented:
Sure, but I might not be able to get back onto my computer until about an hour-
Also, I'm not really familiar with this project or its stack but it looks to me like the code that enables using the generated Toolbox scripts is only in the `win32.ts` file, which, I would assume, is only used on Windows.

tsvetilian-ty commented:
No worries about the video, I just want to see how it behaves on your side. Here is a video of my VM
![fleet-anim](https://user-images.githubusercontent.com/9341546/207462389-171c77c0-aea6-44d3-b658-b4590fd31a16.gif)

Yes, `win32.ts` is Windows only, but that is where the issue with JetBrains Toolbox is. We use the registry to detect an editor on Windows, and IDEs installed through JetBrains Toolbox don't have registry entries, so they are invisible to GitHub  Desktop.

valentinegb commented:
Hmm, not quite what happens for me.

https://user-images.githubusercontent.com/35977727/207465886-771d64e2-d858-4316-8aa9-bcdfe22c5bb8.mov

*Edited so the video works better*

tsvetilian-ty commented:
It is slow because it's on VM, and it's a gif (15fps). 
Just a suggestion that might be why you have issues when you checkout the branch. It's a good idea to run `yarn build:dev` (or `npm run build:dev`) to ensure you have the latest state of the build cached.

valentinegb commented:
No I meant the video *I* sent- and I ran `yarn build:dev` before I started it, if that's what you mean.

tsvetilian-ty commented:
I couldn't see the video at all, so I had to open it on mobile 😆 Perhaps you could try to run GitHub Desktop from external terminal 🤔

valentinegb commented:
Oh haha- this is a new video, it should work better.

https://user-images.githubusercontent.com/35977727/207476235-685cdb96-31b2-43d0-b0a3-70336fd1d58b.mp4

So when I ran it in my terminal it did work. I still wonder why it didn't before... but I guess it probably doesn't really matter.

(Enjoy the mini-tour of my files lol)

sergiou87 commented on app/src/lib/editors/win32.ts:
```diff
@@ -60,7 +70,8 @@ type WindowsExternalEditor = {
   readonly publisher: string
-} & WindowsExternalEditorPathInfo
+} & WindowsExternalEditorPathInfo &
+  WindowsExternalEditorJetbrainsToolbox
```
Since this is just a field you added, I don't think we need an additional type to `&` here. `WindowsExternalEditorPathInfo` was different because it has two different types inside.

However, I'm not sure about the new `toolboxShellScriptName` optional field because it's not tied/required for JetBrains IDEs, but I don't see an easy way of getting TypeScript to enforce that.

In any case, I think the field name should be more JetBrains specific, something like `jetBrainsToolboxScriptName`

tsvetilian-ty replied:
Initially, I thought it might require more fields, so I added the type. But in its final form, since it's a single and optional field, I added it to the `WindowsExternalEditor` type and renamed it as you suggested to `jetBrainsToolboxScriptName`

sergiou87 commented on app/src/lib/editors/win32.ts:
```diff
@@ -18,6 +18,16 @@ interface IWindowsAppInformation {
+   * Default shell script name for JetBrains Products
+   * To get the script name go to:
+   * JetBrains Toolbox > Editor settings > Shell script name
+   * Note: Don't use on JetBrains community and edu editions
```
Why  not? 🤔  Probably worth mentioning that

tsvetilian-ty replied:
The main issue is that JetBrains Toolbox generated the same script name for products without considering their edition.
JetBrains PyCharm Example:
1. Install JetBrains PyCharm Community
2. At this point, JetBrains Toolbox will generate a shell script called `pycharm`
3. Install JetBrains PyCharm Professional
4. JetBrains Toolbox will generate a new script with the same name, `pycharm` and will override the script generated for the community version

I added this example to [editor-integration.md](https://github.com/desktop/desktop/pull/15780/files#diff-2a91c8695284b34dee3845d11a7595422c8dd1a773c7b43e31cdd8a89655104c) and more documentation on how the auto-generation process works

sergiou87 commented on app/src/lib/editors/win32.ts:
```diff
@@ -465,6 +493,31 @@ async function findApplication(editor: WindowsExternalEditor) {
     }
   }
 
+  if (editor.toolboxShellScriptName) {
+    const toolboxRegistryReference = [
+      CurrentUserUninstallKey('toolbox'),
+      Wow64LocalMachineUninstallKey('toolbox'),
+    ]
+
+    for (const { key, subKey } of toolboxRegistryReference) {
+      const keys = enumerateValues(key, subKey)
+
+      if (keys.length > 0) {
+        const editorPathInToolbox = Path.join(
+          getKeyOrEmpty(keys, 'UninstallString'),
+          '..',
+          '..',
+          'scripts',
+          `${editor.toolboxShellScriptName}.cmd`
+        )
+        const exists = await pathExists(editorPathInToolbox)
+        if (exists) {
+          return editorPathInToolbox
+        }
+      }
+    }
+  }
+
   return null
 }
```
I think I'd extract this to a function:
```suggestion
  return findJetBrainsToolboxApplication(editor)
}

async function findJetBrainsToolboxApplication(editor: WindowsExternalEditor) {
  if (!editor.toolboxShellScriptName) {
    return null
  }
  
  const toolboxRegistryReference = [
    CurrentUserUninstallKey('toolbox'),
    Wow64LocalMachineUninstallKey('toolbox'),
  ]

  for (const { key, subKey } of toolboxRegistryReference) {
    const keys = enumerateValues(key, subKey)

    if (keys.length > 0) {
      const editorPathInToolbox = Path.join(
        getKeyOrEmpty(keys, 'UninstallString'),
        '..',
        '..',
        'scripts',
        `${editor.toolboxShellScriptName}.cmd`
      )
      const exists = await pathExists(editorPathInToolbox)
      if (exists) {
        return editorPathInToolbox
      }
    }
  }

  return null
}
```

sergiou87 commented on docs/technical/editor-integration.md:
```diff
@@ -203,6 +204,25 @@ location with an interface that doesn't change between updates.
+
+Now GitHub Desktop support editors installed through JetBrains Toolbox. 
+The technique used to achieve that is using `toolboxShellScriptName` parameter
+to check if, in the default section for scripts in JetBrainsm Toolbox a script
```
Nit:
```suggestion
to check if, in the default section for scripts in JetBrainsm Toolbox, a script
```

sergiou87 reviewed (changes requested):
Thank you for this!! ❤️ 

To be honest, my intention was to thoroughly review #14843 but that's a bigger chunk of work. This PR addresses a long-standing issue with JetBrains Toolbox users in a more surgical way  🎉 

The main limitation I see is it only works with the default script names, but I'm not too worried since Desktop support for other external editors is quite rigid anyway.

And a nice to have would be making that new property required/available for JetBrains IDEs but not for the rest, but I'm not sure we can do that in TypeScript without over-engineering a type, so… 🤷‍♂️  Maybe @niik has some smart idea for that 😄

tsvetilian-ty committed:
Remove unnecessary type for toolbox

tsvetilian-ty committed:
Extract JetBrains Toolbox detection to a function

tsvetilian-ty committed:
Update docs

sergiou87 commented on docs/technical/editor-integration.md:
````diff
@@ -206,20 +206,32 @@ missing or lost it won't let you launch the external editor.
 },
 ```
 
+**Note:** Don't use `jetBrainsToolboxScriptName` field on JetBrains community and edu
+editions. When JetBrains Toolbox generates the scripts. It doesn't consider the
````
Small nit:
```suggestion
**Note:** Don't use `jetBrainsToolboxScriptName` field on JetBrains community and edu
editions. When JetBrains Toolbox generates the scripts, it doesn't consider the
```

Also maybe this should be reworded not to explicitly mention the community and edu editions, but to explain that, in general, Desktop can't handle multiple editions of the same JetBrains IDE installed simultaneously, since JetBrains Toolbox will use the same script name and override it for existing versions. Is that correct? Or is it really specific to community and edu editions?

tsvetilian-ty replied:
You're right. Instead of using community and edu editions, I should use a more generic statement, so I changed it. There is nothing specific to the community and edu, and I just used them as examples of product editions. 

Yes, you're correct. GitHub Desktop can't detect multiple editions because a new installation overrides the existing script.

tsvetilian-ty committed:
Update the usage docs for JetBrains Toolbox

sergiou87 reviewed (approved):
Thank you for those changes!! I tested it and works as expected 👏
